### PR TITLE
Dualstack interop testing enablement

### PR DESCRIPTION
**Pull request by larry-safran**
Fix flag string for enabling dual stack and give server ability to parse argument `address_type`

### CODE REVIEW

File: interop-testing/src/main/java/io/grpc/testing/integration/TestServiceServer.java [hunk 1]
```diff
@@ -130,11 +138,14 @@ void parseArgs(String[] args) {
               + "\n  --local_handshaker_port=PORT"
               + "\n                        Use local ALTS handshaker service on the specified port "
               + "\n                        for testing. Only effective when --use_alts=true."
+              + "\n  --address_type=IPV4|IPV6|IPV4_IPV6"
```
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/TestServiceServer.java:141]: Will there be an update to https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md ?

File: interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java [hunk 2]
```diff
@@ -162,11 +174,15 @@ private void parseArgs(String[] args) {
               + s.secureMode
               + "\n  --server_id=STRING  server ID for response."
               + "\n                      Default: "
-              + s.serverId);
+              + s.serverId
+              + "\n  --address_type=STRING  type of IP address to bind to (IPV4|IPV6|IPV4_IPV6)."
```
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java:178]: Will there be an update to https://github.com/grpc/grpc/blob/master/doc/xds-test-descriptions.md ?

File: interop-testing/src/test/java/io/grpc/testing/integration/StressTestClientTest.java [hunk 3]
```diff
@@ -108,58 +111,112 @@ public void allCommandlineSwitchesAreSupported() {
     assertEquals(60, client.metricsLogRateSecs());
   }
 
+  @Test
+  public void gaugesShouldBeExported_ipv4() throws Exception {
+    checkGaugesShouldBeExported(IPV4);
```
Comment (ejona86) [inline interop-testing/src/test/java/io/grpc/testing/integration/StressTestClientTest.java:116]: :frowning_face:, do we really need to check the gauges for all the different IP binding methods?

Is there any document describing why we are messing with the stress client?

Comment (larry-safran) [reply interop-testing/src/test/java/io/grpc/testing/integration/StressTestClientTest.java:116]: Because it is the only test that actually talks to a server over IP.

Comment (ejona86) [reply interop-testing/src/test/java/io/grpc/testing/integration/StressTestClientTest.java:116]: So make a new test? I don't follow. It doesn't even need a client; you could just open a TCP connection to see if the port is open, if you wanted.

What clued me in to this is the switch statement. We really want to avoid logic in tests.

Comment (larry-safran) [reply interop-testing/src/test/java/io/grpc/testing/integration/StressTestClientTest.java]: done

**Comment by larry-safran** [comment]
@ejona86 ping

File: interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java [hunk 4]
```diff
@@ -138,6 +147,10 @@ private void parseArgs(String[] args) {
         enableCsmObservability = Boolean.valueOf(value);
       } else if ("server_id".equals(key)) {
         serverId = value;
+      } else if ("address_type".equals(key)) {
+        if (!value.isEmpty()) {
```
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java:151]: If the value is empty, we want to fail. What is the benefit of silently ignoring it?

File: interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java [hunk 5]
```diff
@@ -216,8 +233,33 @@ private void start() throws Exception {
               .build();
       server.start();
     } else {
+      ServerBuilder<?> serverBuilder;
+      ServerCredentials insecureServerCreds = InsecureServerCredentials.create();
+      switch (addressType) {
```
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java:238]: Add a failure in the `if (secureMode)` block to fail if `addressType != IPV4_IPV6`?

File: xds/src/main/java/io/grpc/xds/XdsEndpointResource.java [hunk 6]
```diff
@@ -101,7 +100,8 @@ protected EdsUpdate doParse(Args args, Message unpackedMessage) throws ResourceI
   }
 
   private static boolean isEnabledXdsDualStack() {
-    return GrpcUtil.getFlag(GRPC_EXPERIMENTAL_XDS_DUALSTACK_ENDPOINTS, false);
+    return GrpcUtil.getFlag(PickFirstLoadBalancerProvider.GRPC_EXPERIMENTAL_XDS_DUALSTACK_ENDPOINTS,
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsEndpointResource.java:103]: We will need to be careful here to not flip the flag before it is passing interop. We can flip the happy eyeballs flag without interop passing, but we shouldn't flip this one.

File: interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    try {
+      creds = TlsChannelCredentials.newBuilder()
+          .trustManager(TlsTesting.loadCert("ca.pem"))
+          .build();
+    } catch (Exception ex) {
+      throw new RuntimeException(ex);
+    }
+
+    ManagedChannelBuilder<?> builder;
+    if (port == 0) {
```
Comment (ejona86) [inline interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java:101]: Delete this case. It isn't used.

File: interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    String target = getTargetServer(addressType);
+
+    int port = server.getPort();
+    try {
+      ManagedChannel realChannel = createChannel(port, target);
+      Channel channel = cleanupRule.register(realChannel);
+      TestServiceGrpc.TestServiceBlockingStub stub = TestServiceGrpc.newBlockingStub(channel);
+
+      assertEquals(EMPTY, stub.emptyCall(EMPTY));
+    } catch (Exception e) {
```
Comment (ejona86) [inline interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java:83]: Delete this catch. It doesn't do anything. (The method already has `throws Exception`)

File: interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+import io.grpc.TlsChannelCredentials;
+import io.grpc.testing.GrpcCleanupRule;
+import io.grpc.testing.TlsTesting;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link StressTestClient}. */
```
Comment (ejona86) [inline interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java:38]: I assume this is wrong.

File: interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+@RunWith(JUnit4.class)
+public class DualStackClientTest {
```
Comment (ejona86) [inline interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java:40]: I'm confused what this is trying to test. What is the goal here? `checkNoAddressType()` in particular existing really confuses me.

File: interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+  protected static final EmptyProtos.Empty EMPTY = EmptyProtos.Empty.getDefaultInstance();
+
+  @Rule
+  public final GrpcCleanupRule cleanupRule = new GrpcCleanupRule();
+
+  @Rule
+  public final Timeout globalTimeout = Timeout.seconds(10);
```
Comment (ejona86) [inline interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java:47]: This will leak the server. Why not just use a timeout on the unary RPC?

File: interop-testing/src/main/java/io/grpc/testing/integration/Util.java [hunk 12]
```diff
@@ -66,4 +71,32 @@ public static void assertEquals(List<? extends MessageLite> expected,
...
+  }
+
+  static SocketAddress getV4Address(int port) throws UnknownHostException {
+    InetAddress[] addresses = InetAddress.getAllByName(InetAddress.getLocalHost().getHostName());
+    for (InetAddress address : addresses) {
+      if (address.getAddress().length == 4) {
+        return new java.net.InetSocketAddress(address, port);
+      }
+    }
+    return null; // should never happen
```
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/Util.java:93]: This can totally 100% happen. In fact, it is happening with the Mac machine now.

File: interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+  @Test
+  public void checkNoAddressType() throws Exception {
+    checkConnectionWorks(null);
+  }
+
+  private void checkConnectionWorks(Util.AddressType addressType)
+      throws Exception {
+
+    TestServiceServer server = getAndStartTestServiceServer(addressType);
+    String target = getTargetServer(addressType);
```
Comment (ejona86) [inline interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java:74]: How does changing the target string between the tests prove anything? It seems we should use `localhost` unconditionally here, expecting localhost maps to `::1` and `127.0.0.1`. We can have an Assume check that ignores these tests if it doesn't map to at least one IPv4 and one IPv6?

File: interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+      assertEquals(EMPTY, stub.emptyCall(EMPTY));
+    } catch (Exception e) {
+      throw new AssertionError(e);
+    } finally {
+      server.stop();
+    }
+  }
+
+  private static ManagedChannel createChannel(int port, String target) {
```
Comment (ejona86) [inline interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java:90]: `throws Exception` to remove the unhelpful catch+rethrow?

File: interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    builder.overrideAuthority("foo.test.google.fr");
+    return builder.build();
+  }
+
+  private static TestServiceServer getAndStartTestServiceServer(Util.AddressType addressType)
+      throws Exception {
+    TestServiceServer server = new TestServiceServer();
+    String[] args =
+        addressType != null
+        ? new String[]{"--port=8082", "--use_tls=true", "--address_type=" + addressType}
```
Comment (ejona86) [inline interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java:116]: I'm wary of having hard-coded ports in tests. They are flaky. I'm not sure about your goals, so I'm not sure how we can address it. It does look like we can't use port `0`.

File: interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java (hunk 4 above)
Comment (larry-safran) [reply interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java:151]: If they don't put anything it will fail.  if they explicitly put empty string as an argument then they get the default.

Comment (larry-safran) [reply interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java:151]: But I don't really care, so will remove it.

Comment (ejona86) [reply interop-testing/src/main/java/io/grpc/testing/integration/XdsTestServer.java:151]: They didn't get the default. They got a noop. If they had `--address_type=IPV6 --address_type=` then they'd get IPV6.

File: interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java (hunk 13 above)
Comment (ejona86) [reply interop-testing/src/test/java/io/grpc/testing/integration/DualStackClientTest.java]: Discussed offline. We are testing the server, not the client (thus the test class rename). Thus we want different ips.

**Review by ejona86 — APPROVED** [review]

File: interop-testing/src/main/java/io/grpc/testing/integration/StressTestClient.java [hunk 16]
```diff
@@ -308,7 +308,9 @@ void shutdown() {
     }
 
     try {
-      metricsServer.shutdownNow();
+      if (metricsServer != null) {
+        metricsServer.shutdownNow();
+      }
```
Comment (sergiitk) [inline interop-testing/src/main/java/io/grpc/testing/integration/StressTestClient.java:313]: Since this is unrelated to the change, could we please move it to a separate PR? To avoid any confusion, and in case we need to roll back this PR.

Comment (sergiitk) [inline interop-testing/src/main/java/io/grpc/testing/integration/StressTestClient.java:313]: Since this is unrelated to the change, could we please move it to a separate PR? To avoid any confusion, and in case we need to roll back this PR.

File: interop-testing/src/test/java/io/grpc/testing/integration/XdsTestServerTest.java [hunk 17]
```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2016 The gRPC Authors
```
Comment (sergiitk) [inline interop-testing/src/test/java/io/grpc/testing/integration/XdsTestServerTest.java:2]:
```suggestion
 * Copyright 2024 The gRPC Authors
```

File: core/src/main/java/io/grpc/internal/PickFirstLoadBalancerProvider.java [hunk 18]
```diff
@@ -33,15 +33,15 @@
  * down the address list and sticks to the first that works.
  */
 public final class PickFirstLoadBalancerProvider extends LoadBalancerProvider {
-  public static final String GRPC_EXPERIMENTAL_XDS_DUALSTACK_ENDPOINTS =
-      "GRPC_EXPERIMENTAL_XDS_DUALSTACK_ENDPOINTS";
+  public static final String GRPC_PF_USE_HAPPY_EYEBALLS = "GRPC_PF_USE_HAPPY_EYEBALLS";
```
Comment (sergiitk) [inline core/src/main/java/io/grpc/internal/PickFirstLoadBalancerProvider.java:36]: @ejona86 - does it make sense to separate PSM Interop server changes from the dualstack enablement env vars? I'd prefer that, but if not possible/desirable, I'm good with this too.

**Review by sergiitk — CHANGES_REQUESTED** [review]
Before merging, let's confirm PSM Interop passes with the changes to the server.

**Comment by sergiitk** [comment]
- [ ] [grpc/java/master/branch/psm-security](https://source.cloud.google.com/results/invocations/4033b52b-555b-4782-8cfd-a0143fb81a65)
- [ ] [grpc/java/master/branch/xds_k8s_lb](https://source.cloud.google.com/results/invocations/6a64dfa0-f4ef-49cb-9eb7-26aff8e5ffb5)
- [ ] [grpc/java/master/branch/xds_url_map](https://source.cloud.google.com/results/invocations/dcb56597-aaee-43b9-b884-666f13bda421)
- [ ] [grpc/java/master/branch/psm-csm](https://source.cloud.google.com/results/invocations/e054f5b4-60be-44b3-965a-353b72cad0b1)

File: xds/src/main/java/io/grpc/xds/XdsEndpointResource.java [hunk 19]
```diff
@@ -49,8 +49,9 @@
 class XdsEndpointResource extends XdsResourceType<EdsUpdate> {
   static final String ADS_TYPE_URL_EDS =
       "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment";
-  static final String GRPC_EXPERIMENTAL_XDS_DUALSTACK_ENDPOINTS =
-      "grpc.experimental.xdsDualstackEndpoints";
+
+  public static final String GRPC_EXPERIMENTAL_XDS_DUALSTACK_ENDPOINTS =
+      "GRPC_EXPERIMENTAL_HAPPY_EYEBALLS";
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsEndpointResource.java:54]: HAPPY_EYEBALLS? This should be `"GRPC_EXPERIMENTAL_XDS_DUALSTACK_ENDPOINTS"`, per the gRFC, right?

Comment (larry-safran) [reply xds/src/main/java/io/grpc/xds/XdsEndpointResource.java]: Fixed

**Review by sergiitk — APPROVED** [review]
